Ambiguous dependencies for type java.util.concurrent.ExecutorService and qualifiers [@Default] when using SmallRye Context Propagation

**Describe the bug**
Having a bean injecting an ExecutorService `@Inject ExecutorService executorService`, an Ambiguous dependencies error is thrown:
```
[ERROR] 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.AmbiguousResolutionException: Ambiguous dependencies for type java.util.concurrent.ExecutorService and qualifiers [@Default]
[ERROR] 	- java member: org.acme.quickstart.GreetingResource#executorService
[ERROR] 	- declared on CLASS bean [types=[org.acme.quickstart.GreetingResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.acme.quickstart.GreetingResource]
[ERROR] 	- available beans:
[ERROR] 		- PRODUCER METHOD bean [types=[io.netty.channel.EventLoopGroup, java.util.concurrent.Executor, java.util.concurrent.ExecutorService, java.lang.Iterable<io.netty.util.concurrent.EventExecutor>, java.lang.Object, io.netty.util.concurrent.EventExecutorGroup, java.util.concurrent.ScheduledExecutorService], qualifiers=[@Default, @Any], target=io.netty.channel.EventLoopGroup produce_io_netty_channel_EventLoopGroup_bf21a9e8fbc5a3846fb05b4fa0859e0917b2202f(), declaringBean=io.quarkus.arc.runtimebean.RuntimeBeanProducers]
[ERROR] 		- PRODUCER METHOD bean [types=[java.util.concurrent.Executor, org.eclipse.microprofile.context.ManagedExecutor, java.util.concurrent.ExecutorService, java.lang.Object], qualifiers=[@Default, @Any], target=org.eclipse.microprofile.context.ManagedExecutor getAllManagedExecutor(), declaringBean=io.quarkus.smallrye.context.runtime.SmallRyeContextPropagationProvider]
```
**Expected behavior**
The ExecutorService produced by the Context Propagation extension is used (it works in 0.21.1)

**Actual behavior**
The build fails with the stacktrace above

**To Reproduce**
Steps to reproduce the behavior:
1. `mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=getting-started \
    -DclassName="org.acme.quickstart.GreetingResource" \
    -Dpath="/hello" -Dextensions="context-propagation"` 
2. Add a `@Inject ExecutorService executorService` in the `GreetingResource` class
3. `mvn clean package -DskipTests`

**Configuration**
```properties
# Add your application.properties here, if applicable.

```

**Screenshots**
(If applicable, add screenshots to help explain your problem.)

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: `Linux ggastald-laptop 5.2.9-200.fc30.x86_64 #1 SMP Fri Aug 16 21:37:45 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
 - Output of `java -version`: 
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_212-b03)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.212-b03, mixed mode)

 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 999-SNAPSHOT (master)

**Additional context**
Does not happen in 0.21.1

--- a/extensions/netty/deployment/src/main/java/io/quarkus/netty/deployment/EventLoopSupplierBuildItem.java
+++ b/extensions/netty/deployment/src/main/java/io/quarkus/netty/deployment/EventLoopSupplierBuildItem.java
@@ -1,0 +1,25 @@
+package io.quarkus.netty.deployment;
+
+import java.util.function.Supplier;
+
+import io.netty.channel.EventLoopGroup;
+import io.quarkus.builder.item.SimpleBuildItem;
+
+public class EventLoopSupplierBuildItem extends SimpleBuildItem {
+
+    private final Supplier<EventLoopGroup> mainSupplier;
+    private final Supplier<EventLoopGroup> bossSupplier;
+
+    public EventLoopSupplierBuildItem(Supplier<EventLoopGroup> mainSupplier, Supplier<EventLoopGroup> bossSupplier) {
+        this.mainSupplier = mainSupplier;
+        this.bossSupplier = bossSupplier;
+    }
+
+    public Supplier<EventLoopGroup> getMainSupplier() {
+        return mainSupplier;
+    }
+
+    public Supplier<EventLoopGroup> getBossSupplier() {
+        return bossSupplier;
+    }
+}

--- a/extensions/netty/deployment/src/main/java/io/quarkus/netty/deployment/NettyProcessor.java
+++ b/extensions/netty/deployment/src/main/java/io/quarkus/netty/deployment/NettyProcessor.java
@@ -1,5 +1,6 @@
 package io.quarkus.netty.deployment;
 
+import java.util.Optional;
 import java.util.function.Supplier;
 
 import javax.enterprise.context.ApplicationScoped;
@@ -16,7 +17,8 @@ import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.JniBuildItem;
 import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.substrate.SubstrateConfigBuildItem;
-import io.quarkus.netty.BossGroup;
+import io.quarkus.netty.BossEventLoopGroup;
+import io.quarkus.netty.MainEventLoopGroup;
 import io.quarkus.netty.runtime.NettyRecorder;
 
 class NettyProcessor {
@@ -99,19 +101,27 @@ class NettyProcessor {
     @BuildStep
     @Record(ExecutionTime.STATIC_INIT)
     void createExecutors(BuildProducer<RuntimeBeanBuildItem> runtimeBeanBuildItemBuildProducer,
+            Optional<EventLoopSupplierBuildItem> loopSupplierBuildItem,
             NettyRecorder recorder) {
         //TODO: configuration
-        Supplier<Object> boss = recorder.createEventLoop(1);
-        Supplier<Object> worker = recorder.createEventLoop(0);
-
+        Supplier<Object> boss;
+        Supplier<Object> main;
+        if (loopSupplierBuildItem.isPresent()) {
+            boss = (Supplier) loopSupplierBuildItem.get().getBossSupplier();
+            main = (Supplier) loopSupplierBuildItem.get().getMainSupplier();
+        } else {
+            boss = recorder.createEventLoop(1);
+            main = recorder.createEventLoop(0);
+        }
         runtimeBeanBuildItemBuildProducer.produce(RuntimeBeanBuildItem.builder(EventLoopGroup.class)
                 .setSupplier(boss)
                 .setScope(ApplicationScoped.class)
-                .addQualifier(BossGroup.class)
+                .addQualifier(BossEventLoopGroup.class)
                 .build());
         runtimeBeanBuildItemBuildProducer.produce(RuntimeBeanBuildItem.builder(EventLoopGroup.class)
-                .setSupplier(worker)
+                .setSupplier(main)
                 .setScope(ApplicationScoped.class)
+                .addQualifier(MainEventLoopGroup.class)
                 .build());
     }
 

--- a/extensions/netty/runtime/src/main/java/io/quarkus/netty/BossEventLoopGroup.java
+++ b/extensions/netty/runtime/src/main/java/io/quarkus/netty/BossEventLoopGroup.java
@@ -7,5 +7,5 @@ import javax.inject.Qualifier;
 
 @Qualifier
 @Retention(RetentionPolicy.RUNTIME)
-public @interface BossGroup {
+public @interface BossEventLoopGroup {
 }

--- a/extensions/netty/runtime/src/main/java/io/quarkus/netty/MainEventLoopGroup.java
+++ b/extensions/netty/runtime/src/main/java/io/quarkus/netty/MainEventLoopGroup.java
@@ -1,0 +1,11 @@
+package io.quarkus.netty;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+import javax.inject.Qualifier;
+
+@Qualifier
+@Retention(RetentionPolicy.RUNTIME)
+public @interface MainEventLoopGroup {
+}

--- a/extensions/vertx/deployment/src/main/java/io/quarkus/vertx/deployment/VertxProcessor.java
+++ b/extensions/vertx/deployment/src/main/java/io/quarkus/vertx/deployment/VertxProcessor.java
@@ -64,6 +64,7 @@ import io.quarkus.gizmo.MethodCreator;
 import io.quarkus.gizmo.MethodDescriptor;
 import io.quarkus.gizmo.ResultHandle;
 import io.quarkus.gizmo.TryBlock;
+import io.quarkus.netty.deployment.EventLoopSupplierBuildItem;
 import io.quarkus.runtime.RuntimeValue;
 import io.quarkus.vertx.ConsumeEvent;
 import io.quarkus.vertx.runtime.EventConsumerInvoker;
@@ -129,6 +130,11 @@ class VertxProcessor {
     @BuildStep
     AdditionalBeanBuildItem registerBean() {
         return AdditionalBeanBuildItem.unremovableOf(VertxProducer.class);
+    }
+
+    @Record(ExecutionTime.STATIC_INIT)
+    EventLoopSupplierBuildItem eventLoop(VertxRecorder recorder) {
+        return new EventLoopSupplierBuildItem(recorder.bossSupplier(), recorder.mainSupplier());
     }
 
     @BuildStep

--- a/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxRecorder.java
+++ b/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxRecorder.java
@@ -8,11 +8,13 @@ import java.util.*;
 import java.util.Map.Entry;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 import org.jboss.logging.Logger;
 
+import io.netty.channel.EventLoopGroup;
 import io.quarkus.arc.runtime.BeanContainer;
 import io.quarkus.runtime.LaunchMode;
 import io.quarkus.runtime.RuntimeValue;
@@ -27,6 +29,7 @@ import io.vertx.core.eventbus.MessageCodec;
 import io.vertx.core.eventbus.MessageConsumer;
 import io.vertx.core.file.FileSystemOptions;
 import io.vertx.core.http.ClientAuth;
+import io.vertx.core.impl.VertxImpl;
 import io.vertx.core.net.JksOptions;
 import io.vertx.core.net.PemKeyCertOptions;
 import io.vertx.core.net.PemTrustOptions;
@@ -349,5 +352,23 @@ public class VertxRecorder {
     private void registerCodec(Class<?> typeToAdd, MessageCodec codec) {
         EventBus eventBus = vertx.eventBus();
         eventBus.registerDefaultCodec(typeToAdd, codec);
+    }
+
+    public Supplier<EventLoopGroup> bossSupplier() {
+        return new Supplier<EventLoopGroup>() {
+            @Override
+            public EventLoopGroup get() {
+                return ((VertxImpl) vertx).getAcceptorEventLoopGroup();
+            }
+        };
+    }
+
+    public Supplier<EventLoopGroup> mainSupplier() {
+        return new Supplier<EventLoopGroup>() {
+            @Override
+            public EventLoopGroup get() {
+                return vertx.nettyEventLoopGroup();
+            }
+        };
     }
 }

--- a/integration-tests/vertx/src/main/java/io/quarkus/it/vertx/NettyEventLoopResource.java
+++ b/integration-tests/vertx/src/main/java/io/quarkus/it/vertx/NettyEventLoopResource.java
@@ -5,16 +5,18 @@ import javax.ws.rs.GET;
 import javax.ws.rs.Path;
 
 import io.netty.channel.EventLoopGroup;
-import io.quarkus.netty.BossGroup;
+import io.quarkus.netty.BossEventLoopGroup;
+import io.quarkus.netty.MainEventLoopGroup;
 
 @Path("/eventloop")
 public class NettyEventLoopResource {
 
     @Inject
+    @MainEventLoopGroup
     EventLoopGroup worker;
 
     @Inject
-    @BossGroup
+    @BossEventLoopGroup
     EventLoopGroup boss;
 
     @GET

--- a/integration-tests/vertx/src/test/java/io/quarkus/it/vertx/NettyMainEventLoopGroupResourceTest.java
+++ b/integration-tests/vertx/src/test/java/io/quarkus/it/vertx/NettyMainEventLoopGroupResourceTest.java
@@ -8,7 +8,7 @@ import io.quarkus.test.junit.QuarkusTest;
 import io.restassured.RestAssured;
 
 @QuarkusTest
-public class NettyEventLoopGroupResourceTest {
+public class NettyMainEventLoopGroupResourceTest {
 
     @Test
     public void testInjection() {

